tests/late_joiners.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+import os
+import getpass
+import time
+import logging
+import multiprocessing
+import shutil
+from random import seed
+import infra.ccf
+import infra.proc
+import infra.notification
+import infra.net
+import e2e_args
+from threading import Timer
+import random
+import contextlib
+
+from loguru import logger as LOG
+
+# 256 is the number of most recent messages that PBFT keeps in memory before needing to replay the ledger
+TOTAL_REQUESTS = 56
+
+
+def timeout(node, suspend, election_timeout):
+    if suspend:
+        # We want to suspend the nodes' process so we need to initiate a new timer to wake it up eventually
+        node.suspend()
+        next_timeout = random.uniform(2 * election_timeout, 3 * election_timeout)
+        LOG.info(f"New timer set for node {node.node_id} is {next_timeout} seconds")
+        t = Timer(next_timeout, timeout, args=[node, False, 0])
+        t.start()
+    else:
+        node.resume()
+
+
+def run(args):
```

Comment:
I believe we should break this down into smaller "chainable" functions so that it can be run as part of the test suite.

Reply:
This PR has been open for a while and I have already started implementing rollback on a branch off of this. I agree it should be cleaned up/broken into chainable functions, but I would rather do it in the next PR if that's ok.